pkg/vault/contracts/VaultAdmin.sol

```diff
@@ -472,16 +474,30 @@ contract VaultAdmin is IVaultAdmin, VaultCommon, Authentication, VaultGuard {
         _takeDebt(IERC20(underlyingToken), amountUnderlyingRaw);
         _takeDebt(wrappedToken, amountWrappedRaw);
 
-        // Update buffer balances
         bytes32 bufferBalances = _bufferTokenBalances[wrappedToken];
+
+        // The buffer invariant is the sum of buffer token balances converted to underlying.
+        uint256 bufferInvariant = bufferBalances.getBalanceRaw() +
+            wrappedToken.convertToAssets(bufferBalances.getBalanceDerived());
+        // The buffer share rate (equivalent to pool's "BPT rate") is the ratio of the total shares
+        // (equivalent to pool's "total supply") to the buffer invariant. As the rate goes up (i.e., wrapped tokens
+        // become more valuable), the share rate will go down, and the user will receive fewer shares.
+        // The buffer must have been initialized to get here, so the balances cannot be zero, and the rate calculation
+        // will never revert.
+        uint256 currentBufferShareRate = _bufferTotalShares[wrappedToken].divDown(bufferInvariant);
+
+        uint256 bufferInvariantDelta = wrappedToken.convertToAssets(amountWrappedRaw) + amountUnderlyingRaw;
+        // The amount of shares to issue is the added liquidity in underlying terms multiplied by the share rate.
+        issuedShares = bufferInvariantDelta.mulDown(currentBufferShareRate);
```

Comment:
So, this is basically `issuedShares = bufferTotalShares * (invariantRatio - 1)`, where `invariantRatio = (bufferInvariant + bufferInvariantDelta) / bufferInvariant`.

Replace and simplify, and you get 
```
issuedShares = bufferTotalShares * bufferInvariantDelta / currentInvariant
```

I think you can do that with raw multiplication and division without FP math, since it's a product and a division rounded down. The current code uses two FP operations instead, and divides first, multiplies second.